File: src/haruhiloop_cli/policy.py
from __future__ import annotations

import random
from typing import Protocol

from haruhiloop_cli.models import GameState, Scene, SceneChoice, StepCommand, StepRecord


class Policy(Protocol):
    def choose_command(
        self, state: GameState, scenes: list[Scene], choice_map: dict[str, list[SceneChoice]], history: list[StepRecord]
    ) -> StepCommand: ...


class RandomPolicy:
    def choose_command(
        self, state: GameState, scenes: list[Scene], choice_map: dict[str, list[SceneChoice]], history: list[StepRecord]
    ) -> StepCommand:
        _ = state, history
        scene = random.choice(scenes)
        choice = random.choice(choice_map[scene.scene_id])
        return StepCommand(scene_id=scene.scene_id, choice_id=choice.choice_id)


class GreedyPolicy:
    def choose_command(
        self, state: GameState, scenes: list[Scene], choice_map: dict[str, list[SceneChoice]], history: list[StepRecord]
    ) -> StepCommand:
        _ = history

        def score(choice: SceneChoice) -> int:
            value = (
                2 * choice.delta_satisfaction
                + 2 * choice.delta_stability
                + 3 * choice.delta_clue_points
            )
            if state.stability < 40 and ("coordination" in choice.tags or "nagato_relief" in choice.tags):
                value += 8
            if state.satisfaction < 45 and ("festival" in choice.tags or "haruhi_route" in choice.tags):
                value += 8
            if state.clue_points > 8 and "truth_sync" in choice.tags:
                value += 6
            return value

        all_choices: list[SceneChoice] = []
        for scene in scenes:
            all_choices.extend(choice_map.get(scene.scene_id, []))
        best = max(all_choices, key=score)
        return StepCommand(scene_id=best.scene_id, choice_id=best.choice_id)


def create_policy(policy_name: str) -> Policy:
    if policy_name == "random":
        return RandomPolicy()
    if policy_name == "greedy":
        return GreedyPolicy()
    raise ValueError(f"未知策略：{policy_name}")

File: web/src/application/policy.ts
// 与 src/haruhiloop_cli/policy.py 1:1 对齐：random / greedy 两策略。

import { GameState, Scene, SceneChoice, StepCommand, StepRecord } from "../domain/models";
import { mulberry32 } from "../infrastructure/prng";

export interface Policy {
  chooseCommand(
    state: GameState,
    scenes: Scene[],
    choiceMap: Record<string, SceneChoice[]>,
    history: StepRecord[],
  ): StepCommand;
}

export class RandomPolicy implements Policy {
  private readonly rng: () => number;
  constructor(seed: number | null = null) {
    const seedNum = seed ?? ((Date.now() & 0xffffffff) >>> 0);
    this.rng = mulberry32(seedNum);
  }
  chooseCommand(
    _state: GameState,
    scenes: Scene[],
    choiceMap: Record<string, SceneChoice[]>,
    _history: StepRecord[],
  ): StepCommand {
    const scene = scenes[Math.floor(this.rng() * scenes.length)] ?? scenes[0]!;
    const choices = choiceMap[scene.scene_id] ?? [];
    const choice = choices[Math.floor(this.rng() * choices.length)] ?? choices[0]!;
    return { scene_id: scene.scene_id, choice_id: choice.choice_id };
  }
}

export class GreedyPolicy implements Policy {
  chooseCommand(
    state: GameState,
    scenes: Scene[],
    choiceMap: Record<string, SceneChoice[]>,
    _history: StepRecord[],
  ): StepCommand {
    const score = (choice: SceneChoice): number => {
      let value =
        2 * choice.delta_satisfaction +
        2 * choice.delta_stability +
        3 * choice.delta_clue_points;
      if (
        state.stability < 40 &&
        (choice.tags.includes("coordination") || choice.tags.includes("nagato_relief"))
      ) {
        value += 8;
      }
      if (
        state.satisfaction < 45 &&
        (choice.tags.includes("festival") || choice.tags.includes("haruhi_route"))
      ) {
        value += 8;
      }
      if (state.clue_points > 8 && choice.tags.includes("truth_sync")) value += 6;
      return value;
    };
    const all: SceneChoice[] = [];
    for (const scene of scenes) {
      all.push(...(choiceMap[scene.scene_id] ?? []));
    }
    let best = all[0]!;
    let bestScore = score(best);
    for (const c of all) {
      const s = score(c);
      if (s > bestScore) {
        best = c;
        bestScore = s;
      }
    }
    return { scene_id: best.scene_id, choice_id: best.choice_id };
  }
}

export function createPolicy(name: string, seed: number | null = null): Policy {
  if (name === "random") return new RandomPolicy(seed);
  if (name === "greedy") return new GreedyPolicy();
  throw new Error(`未知策略：${name}`);
}
